src/types/publicTypes.ts

```diff
@@ -1,28 +1,53 @@
 import {
+  CnsSupportedNetwork,
+  EnsSupportedNetwork,
   EventFilter,
   RequestArguments,
   RpcProviderLogEntry,
   TransactionRequest,
-} from './types';
+  ZnsSupportedNetwork,
+} from '.';
 
-/**
- * SourceDefinition object
- * @typedef {Object} SourceDefinition
- * @property {string} [url] - main blockchain api url
- * @property {string | number} [network] - blockchain network
- */
-export interface SourceDefinition {
-  url?: string;
-  network?: string | number;
-  registry?: string;
-  provider?: Provider;
+export type CnsSupportedNetworks = typeof CnsSupportedNetwork.type;
+export type EnsSupportedNetworks = typeof EnsSupportedNetwork.type;
+export type ZnsSupportedNetworks = typeof ZnsSupportedNetwork.type;
+
+export type CnsConfig  = {
+  url?: string,
+  network: CnsSupportedNetworks,
+  proxyReaderAddress?: string
+};
+
+export type Api = { api: true, url?: string };
+
+export type EnsConfig = {
+  url?: string,
+  network: EnsSupportedNetworks,
+  registryAddress?: string
+}
+
+export type ZnsConfig = {
+  url?: string,
+  network: ZnsSupportedNetworks,
+  registryAddress?: string
 }
 
+export type CnsSource = CnsConfig | { provider: Provider, network: CnsSupportedNetworks, proxyReaderAddress?: string };
+export type EnsSource = EnsConfig | { provider: Provider, network: EnsSupportedNetworks, registryAddress?: string };
+export type ZnsSource = ZnsConfig | { provider: Provider, network: ZnsSupportedNetworks, registryAddress?: string };
```

Comment:
I'm still not quite happy that we have to duplicate each Config declaration only because they have either `url` or `provider` field.

I'm curios what you think about having it like this:
```
export type CnsConfig  = {
  // Source can be either a URL to Ethereum JSON RPC or a Provider instance. It'll be an Infura provider by default.
  source?: string | Provider,
  network: CnsSupportedNetworks,
  proxyReaderAddress?: string
};
```

I'm still not happy about having either types at all, but this variant seems a bit more readable. It lets us get rid of `CnsSource` and makes the difference between `CnsConfig` and `{ provider: Provider, network: CnsSupportedNetworks, proxyReaderAddress?: string }` more obvious.